### DISCUSSION TITLE
Raise exception in python-fastApi middleware

### DISCUSSION THREAD

**Discussion by priyankagupta34** [fastapi/fastapi#9053] [Questions] [answered]
I am trying to validate token in fastapi middleware but this seems impossible. As i am thinking middleware needs to make next call although its not required. I am not able to find any good solution to handle token in one go in this python-fastapi backend. Any help is appreciated.

```
@app.middleware("http")
async def add_middleware_here(request: Request, call_next):
    token = request.headers["Authorization"]
    try:
        verification_of_token = verify_token(token)
        if verification_of_token:
            response = await call_next(request)
            return response
    except InvalidSignatureError as er:
        raise HTTPException(status_code=401)
```
My expectation is that it should throw 401 error once wrong token is provided.
Instead it throws 500 with below Exception traceback:
```
Traceback (most recent call last):
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 385, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\fastapi\applications.py", line 140, in __call__
    await super().__call__(scope, receive, send)
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\starlette\applications.py", line 134, in __call__
    await self.error_middleware(scope, receive, send)
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 178, in __call__
    raise exc from None
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 156, in __call__
    await self.app(scope, receive, _send)
  File "c:\users\covered\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\base.py", line 26, in __call__
    await response(scope, receive, send)
TypeError: 'NoneType' object is not callable

```

**Comment by LittleLampLight** [comment]
What kind of behavior do you expect when `if verification_of_token:` is elsed? Seems like you must `raise HTTPException(status_code=401)` then. 

Also there is also option to use https://fastapi.tiangolo.com/tutorial/dependencies/

**Comment by priyankagupta34** [comment]
> What kind of behavior do you expect when `if verification_of_token:` is elsed? Seems like you must `raise HTTPException(status_code=401)` then.
> 
> Also there is also option to use https://fastapi.tiangolo.com/tutorial/dependencies/

Actually verification_of_token can be true or false depending on token's validity. I am trying to check token here itself before it goes any further. Also `InvalidSignatureError` error comes up when token is invalid. Now problem is: this always fails the backend here by going in 500 interval server error other than going in 401 unauthorized error. I want to throw 401 error.

**Comment by priyankagupta34** [comment]
I have updated the above problem with exception traceback

**Comment by LittleLampLight** [comment]
According to [middleware docs](https://fastapi.tiangolo.com/tutorial/middleware/) you must either return response or call_next func. Seems like your middleware is not properly throwing a `InvalidSignatureError` and if block being ignored. So you `return None` from your middleware which causes exception

**Comment by priyankagupta34** [comment]
So how should i then check validity of token now?

**Comment by LittleLampLight** [comment]
I'd go for something likes. Not sure how's your `verify_token` func looks like. But it's how to you how you make it) It doesn't seems like FastAPI issue) 

```

def verify_token(token):
       if check_token(token):
           return True
       else:
           return False


@app.middleware("http")
async def add_middleware_here(request: Request, call_next):
    token = request.headers["Authorization"]
    if verify_token(token):
        response = await call_next(request)
        return response
    else:
        raise HTTPException(status_code=401)
```

**Comment by priyankagupta34** [comment]
> I'd go for something likes. Not sure how's your `verify_token` func looks like. But it's how to you how you make it) It doesn't seems like FastAPI issue)
> 
> ```
> 
> def verify_token(token):
>        if check_token(token):
>            return True
>        else:
>            return False
> 
> 
> @app.middleware("http")
> async def add_middleware_here(request: Request, call_next):
>     token = request.headers["Authorization"]
>     if verify_token(token):
>         response = await call_next(request)
>         return response
>     else:
>         raise HTTPException(status_code=401)
> ```

This totally sounds reasonable but i have already performed this thing and it throws same error as above.

**Comment by LittleLampLight** [comment]
Well, just to be clear I didn’t tested code above. It was just a graft. Also may be that’s because you are raising exception instead of returning Response object. Give it a few tries. Also take a look at Dependency based authorization it looks sexy

**Comment by vancouverwill** [comment]
hey @priyankagupta34 I just came across the same issue, seems weirdly `HTTPException` doesn't work in middle ware. See https://github.com/tiangolo/fastapi/issues/1125

**Comment by chris-allnutt** [comment]
the comment in there has the solution tho
```return Response(content="", status_code=401)```

but I agree this is unclear  would be good to know if this should be in the documentation or if it's a bug @tiangolo and others think should be addressed.   Something like bubbling up if the caught below exception is already of type HTTPException.

**Reply by YuriiMotov** [reply, marked as answer]
The thing is that handling `HTTPException` is done by `ExceptionMiddleware` which is the innermost middleware in stack.
There is a remark about this in Starlette Docs:

> You should only raise HTTPException inside routing or endpoints. Middleware classes should instead just return appropriate responses directly.

https://www.starlette.io/exceptions/#httpexception

So, the solution is to return `Response` instead of raising `HTTPException`. And, as it was pointed out, you should always either return response or raise exception (this will lead to server error).

```python
def verify_token(token):
       if check_token(token):
           return True
       else:
           return False


@app.middleware("http")
async def add_middleware_here(request: Request, call_next):
    token = request.headers["Authorization"]
    if verify_token(token):
        response = await call_next(request)
        return response
    else:
        return Response(content="Invalid token", status_code=401)
```

**Comment by HuiiBuh** [comment]
This is an intended. So the thing you should do is to return a Response with right status code and message.
See #458

**Comment by Luttik** [comment]
It might not be a bug but you might be able to classify it as an anti-feature 😆. I'd still be super interested in having the option to use the exception middleware for this... It would allow for much more consistency in the processes. 

Having a request not being able to authenticate is an exception in my eyes, therefore these requests should trigger the flows that we have built for errors (via middleware). It'd be a shame if we cannot use that in combination with authentication middleware. Just silently returning a 401 code would break our monitoring systems.

**Comment by BackMountainDevil** [comment]
```python

def verify_token(token):
       if check_token(token):
           return True
       else:
           return False

credentials_exception = HTTPException(
    status_code=status.HTTP_401_UNAUTHORIZED,
    detail="Could not validate credentials",
    headers={"WWW-Authenticate": "Bearer"},
)

@app.middleware("http")
async def add_middleware_here(request: Request, call_next):
    token = request.headers["Authorization"]
    if verify_token(token):
        response = await call_next(request)
        return response
    else:
        raise credentials_exception(status_code=401)
```

this sitll got error. `raise` -> `return` still got error. Then I turn HTTPException into `Responce`

```python
credentials_exception = Response(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content="Could not validate credentials",
                headers={"WWW-Authenticate": "Bearer"},
            )
```
